### PR TITLE
ENH: distributions, complex arguments

### PR DESCRIPTION
**Pull request by ev-br**
Closes https://github.com/scipy/scipy/issues/2237. No idea if statsmodels still wants it, but here it is anyway.

Remove hardcoded `float` dtype in methods of distributions, so that one can do complex-step derivatives (where the underlying special functions work with complex arguments, that is):

```
>>> h = 1e-6
>>> x = 1.
>>> (cauchy.cdf(x + h*1j) / h).imag - cauchy.pdf(x)
-2.6478819137309983e-14
```

### CODE REVIEW

**Comment by josef-pkt** [comment]
@ev-br  as far as I can see and remember

using `x.dtype` willl not work if x is integer. We need dtype at least float to store np.nan.

It took me a while initially to figure out some casting bugs when nan (invalid numbers or numbers not in support of the function) where cast silently to zero,  This was because discrete distribution returned integer arrays for some methods initially.

**Comment by ev-br** [comment]
This PR only changes continuous distributions. They all seem to have an implicit cast: `x = (x-loc)*1.0/scale`.

**Comment by josef-pkt** [comment]
This will still be useful for statsmodels 

We try to use the analytical derivatives if we manage, but at least in some parts with use complex step derivative as a default with a try except protection to switch to standard forward step derivatives.

One main limitation is still scipy.special, AFAIR, norm.cdf from special doesn't work with complex numbers but `erf` works.

Thanks @ev-br

**Comment by josef-pkt** [comment]
> This PR only changes continuous distributions. They all seem to have an implicit cast: x = (x-loc)*1.0/scale.

Very good, I didn't see the implicit casting. 
Are there a few unit test with integers, just to make sure?

**Comment by josef-pkt** [comment]
just to understand the casting compare to before:

What happens with smaller float, float16 (?) and float32? Are they implicitly upgraded to float64 or stay in the smaller float dtype?

**Comment by ev-br** [comment]
> What happens with smaller float, float16 (?) and float32? Are they implicitly upgraded to float64 or stay in the smaller float dtype?

Rright. As often happens in stats.distributions, it worked OK with an odd glitch or two :-). 
In master: int args are cast to float64 by multiplication by 1.0, and everything stays float64 from that point on. Half- and single-precision floats are *not* cast by multiplication by 1.0 because of --- I suppose --- this "The standard casting rules ensure that a scalar cannot up-cast an array unless the scalar is of a fundamentally different kind of data", http://docs.scipy.org/doc/numpy/reference/generated/numpy.find_common_type.html.
So half- and single-precision floats stayed with their original precision unless upcasted by `(x-loc)/scale`. This propagated through the private methods (`_pdf`), to be finally upcast to float64 in https://github.com/scipy/scipy/blob/master/scipy/stats/_distn_infrastructure.py#L1599, because `output` is allocated with the typecode 'd'. I only checked this on a 64-bit machine, not sure how this changes for 32 bits. And I'm sure it's numpy version dependent as well.

All in all this mostly worked, esp since I do not think reduced precision floats saw much usage :-). By 'mostly' I mean there were odd fails for half-precision floats in some distributions depending on the implementation of the the private `_cdf`-type methods.   

https://github.com/ev-br/scipy/commit/1e78ee2e5a9912c9e51582acb6c3ea76f801c32d implements an explicit test which loudly complains about downcasting from float64 to float16 on `wrapcauchy.cdf` and a couple of other places.

So in this PR now makes explicit casts and adds tests.

**Comment by rgommers** [comment]
> I only checked this on a 64-bit machine, not sure how this changes for 32 bits.

It's the same on 32-bits. I wouldn't expect differences there. 32 vs 64 bits differences are mainly showing up for integers (because ``int`` is ``int32`` on 32-bit machines) and floating-point precision, but not for float casting behavior.

**Comment by rgommers** [comment]
This still passes all tests without any test noise for me (on 32-bit).

File: scipy/stats/_distn_infrastructure.py [hunk 1]
```diff
@@ -1587,11 +1587,12 @@ def pdf(self, x, *args, **kwds):
         args, loc, scale = self._parse_args(*args, **kwds)
         x, loc, scale = map(asarray, (x, loc, scale))
         args = tuple(map(asarray, args))
-        x = asarray((x-loc)*1.0/scale)
+        dtyp = np.find_common_type([x.dtype, np.float_], [])
```
Comment (rgommers) [inline scipy/stats/_distn_infrastructure.py]: nitpick: I'd write ``np.float64`` instead of ``np.float_`` for clarity

**Comment by rgommers** [comment]
So about half the distributions accept complex numbers, and the other half fail. The user won't know which work and which don't without trying. Is that really a good situation?

File: scipy/stats/_distn_infrastructure.py (hunk 1 above)
Comment (ev-br) [inline scipy/stats/_distn_infrastructure.py]: Happy to do this if it makes it clearer (not to me,  I'm thoroughly confused by what are np.float vs np.float_ vs np.float64 vs python float which is C double or is it?)

**Comment by josef-pkt** [comment]
@rgommers  AFAICS this just cleans up the mixed up dtype handling for less than float64.

For complex it always raised an exception, now it raises an exception when it doesn't work. The unit test exclude distributions, some methods still work with excluded distributions.
_pdf for norm works, cdf fails in special  

(for statsmodels this would mean we can use those distributions for logpdf, but not cases like Probit where we need the norm.cdf in the loglikelihood function.)


actually, currently I get a warning with implicit downcasting in some cases (which returns wrong result), and casting warning shows up only the first time AFAICS

```
>>> cc
array([ 4.+1.j])

>>> stats.t._pdf(cc, 0.5)
array([ 0.01760022-0.00657159j])
>>> stats.t.pdf(cc, 0.5)            # truncated imag part
array([ 0.01760022])
C:\programs\WinPython-64bit-3.4.3.1\python-3.4.3.amd64\lib\site-packages\scipy\stats\_continuous_distns.py:3036: ComplexWarning: Casting complex values to real discards the imaginary part
  Px /= sqrt(r*pi)*(1+(x**2)/r)**((r+1)/2)

>>> stats.t.pdf(cc, cc)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\programs\WinPython-64bit-3.4.3.1\python-3.4.3.amd64\lib\site-packages\scipy\stats\_distn_infrastructure.py", line 1573, in pdf
    place(output, cond, self._pdf(*goodargs) / scale)
  File "C:\programs\WinPython-64bit-3.4.3.1\python-3.4.3.amd64\lib\site-packages\numpy\lib\function_base.py", line 1484, in place
    return _insert(arr, mask, vals)
TypeError: Cannot cast array data from dtype('complex128') to dtype('float64') according to the rule 'safe'
>>> import scipy
>>> scipy.__version__
'0.15.1'
>>> stats.t._pdf(cc, cc)
array([ 0.00193552-0.00680458j])
```

File: scipy/stats/_distn_infrastructure.py (hunk 1 above)
Comment (rgommers) [inline scipy/stats/_distn_infrastructure.py]: ``np.float`` will be deprecated, there was a thread about why it's bad to use it on numpy-discussion recently.

http://docs.scipy.org/doc/numpy/user/basics.types.html says: ``float_ : Shorthand for float64.`` 

Python ``float`` is also 64-bit and IIRC under the hood a C double.

**Comment by rgommers** [comment]
> @rgommers AFAICS this just cleans up the mixed up dtype handling for less than float64.

Yes, that all looks fine. What I was commenting on was the test that explicitly verifies that complex numbers now work, but with ~50 exceptions. If the "new feature" has value then we probably would like to fix the distributions that with this PR still raise an exception on complex input where possible.

Anyway, it's more a clarifying question than a strong objection from me. But a comment giving some context would be good to add to the test at leat.

File: scipy/stats/_distn_infrastructure.py (hunk 1 above)
Comment (ev-br) [inline scipy/stats/_distn_infrastructure.py]: Done.
Thanks for the explanation!

**Comment by ev-br** [comment]
> So about half the distributions accept complex numbers, and the other half fail. The user won't know which work and which don't without trying. Is that really a good situation?

Uhm well. Obviously, I'd prefer that all of them work, but extending 50-odd special functions to complex plane is not something I'm up to anyway  :-). Clearly, some would be more important than others, and there is no expectation that all of them need to be made working. 

All in all, I think that this PR with 50+ exceptions is already an improvement over master where they all fail because of a hardcoded limitation --- provided there is at least one use case :-). 
But again, if there are strong feelings that it's too much, I'd be fine with just closing this.

**Comment by ev-br** [comment]
> _pdf for norm works, cdf fails in special

> ( for statsmodels this would mean we can use those distributions for logpdf, but not cases like Probit where we need the norm.cdf in the loglikelihood function.)

`norm` relies on `ndtr` (which is real-only), but we also have `erf`, which works in the complex plane:
```
In [27]: h=1e-8j

In [28]: (erf(1. + h) / h)
Out[28]: (0.41510749742059477-84270079.294971481j)

In [29]: np.exp(-1.) * 2. / np.sqrt(np.pi)
Out[29]: 0.41510749742059477
```

File: scipy/stats/tests/test_continuous_basic.py [hunk 2]
```diff
@@ -79,6 +80,22 @@
 # distslow are sorted by speed (very slow to slow)
 
 
+# These distributions fail the complex derivative test below.
+# Here 'fail' mean produce wrong results and/or raise exceptions, depending
+# on the implementation details of corresponding special functions.
```
Comment (josef-pkt) [inline scipy/stats/tests/test_continuous_basic.py]: I think mentioning the github issue would be useful, to find the explanation and discussion faster

**Comment by josef-pkt** [comment]
For statsmodels and statsmodels users it's already a useful improvement. Enhancing and cleaning up individual distributions can be left as future task and future PRs. IMO.

example usage: https://github.com/statsmodels/statsmodels/blob/master/statsmodels/miscmodels/tmodel.py#L154
Here I copied the logpdf of scipy. In cases where statsmodels developers or users want to use the scipy distribution logpdf directly in maximum likelihood models, the complex derivatives will be more accurate than finite difference approximation.  (But still less precise than analytical derivatives of logpdf, if those are available.)

(aside in some cases we are still better of using a different implementation when a parameterization specialized for regression works better and is more stable in that case. I looked at different ways of parameterizing the gamma distribution recently.)

File: scipy/stats/tests/test_continuous_basic.py (hunk 2 above)
Comment (ev-br) [inline scipy/stats/tests/test_continuous_basic.py]: done.

**Comment by argriffing** [comment]
@josef-pkt What exactly is going on with complex derivatives?  The derivatives methods I know about are like
 * [finite differences](https://en.wikipedia.org/wiki/Numerical_differentiation#Finite_difference_formulas)
 * [automatic differentiation](https://en.wikipedia.org/wiki/Automatic_differentiation)
 * [symbolic differentiation](https://en.wikipedia.org/wiki/Symbolic_computation)

I guess 'complex derivatives' isn't symbolic differentiation, but is it more like finite differences or is it more like a hack that somehow does single order forward automatic differentiation?

Ideally I think it would be useful to add the [score function](https://en.wikipedia.org/wiki/Score_(statistics)#Definition) to the stats interface. This could also be used internally by the `.fit` methods which are defined as using maximum likelihood estimation.  On the other hand it wouldn't be sufficient for second derivatives which are often used to construct intervals. Perhaps an unsophisticated implementation of the score function could use 'complex derivatives' for distributions whose logpdf implementation supports it, and finite differences for the others.

**Comment by WarrenWeckesser** [comment]
@argriffing  "complex step differentiation" is a method for computing a (real) finite difference derivative approximation with a single function call by passing in a complex argument.  See http://blogs.mathworks.com/cleve/2013/10/14/complex-step-differentiation/

**Comment by josef-pkt** [comment]
@argriffing  there are some references in some (?) issues. Someone (?) pointed them out on the scipy mailing list, I implemented them for statsmodels and we use it as high precision default when they work. accuracy in examples that I tried is around 1e-12 to 1e-14, but a bit slower than finite differences. 
We never had to tweak the stepsize for them, but they only work for first differences (in the standard implementation)

**Comment by josef-pkt** [comment]
(more advertising: AFAIR, we use them in ARMA and statespace model with the Kalman filter. The cython implementation allows complex variables specifically so we can use complex step derivatives.)

**Comment by argriffing** [comment]
Would it be enough to hide these details within per-distribution 'score' functions? In other words, the user would not have to care about imaginary numbers, but for distributions which support complex step derivatives the users would get the benefit automatically when they call the `score` member function?

**Comment by josef-pkt** [comment]
Hiding them inside the score function (deriv_logpdf for example) is only of limited use. 

First in some cases we need the cdf (like in censored/truncated/discretized models) and maybe the ppf (when we include quantile based estimation.

Second, we currently use numerical derivatives almost exclusively on the entire model, and it would be the simple default for user defined likelihood functions. We use explicit chain rule in several places for analytical derivatives, but coding chainrule also into numerical derivatives requires work.

We use numerical derivatives inside an analytical chain rule in a few places, and I expect we will extend that, but getting to the required modularity will take time.
This PR extends the range where complex step derivatives can be used without additional effort by us or users.

**Comment by ev-br** [comment]
Re complex-step derivatives:
```
In [24]: from scipy.optimize._numdiff import approx_derivative

In [25]: approx_derivative(np.sin, 0.5, method='cs')
Out[25]: array([ 0.87758256])
```

**Comment by argriffing** [comment]
> But again, if there are strong feelings that it's too much, I'd be fine with just closing this.

I don't have any objection to this PR.  It's clear to me that the long term solution for derivatives for statistical analysis is a mix of analytic and automatic differentiation, but this looks like an OK temporary solution.

**Comment by rgommers** [comment]
> > But again, if there are strong feelings that it's too much, I'd be fine with just closing this.

> I don't have any objection to this PR. It's clear to me that the long term solution for derivatives for statistical analysis is a mix of analytic and automatic differentiation, but this looks like an OK temporary solution.

Good discussion, and it's clear that this has value to at least ``statsmodels``. So no objection here either.

**Comment by datnamer** [comment]
Regarding autodiff of numpy code: https://github.com/HIPS/autograd

**Comment by argriffing** [comment]
Merged, thanks @ev-br!

> Regarding autodiff of numpy code: https://github.com/HIPS/autograd

@datnamer Great! I hope some day scipy will include/vendor/depend a package like that. I'm glad to see that automatic differentiation is starting to become almost expected in scientific computing environments, so that if you write a differentiable function then you can take its derivatives for granted.

**Comment by ev-br** [comment]
> Regarding autodiff of numpy code: https://github.com/HIPS/autograd

Would be good to include the link to the topical software list.

**Comment by rgommers** [comment]
@ev-br good suggestion, done in https://github.com/scipy/scipy.org/pull/120

**Pull request merged**